src/main/java/com/coders/repository/ProductRepository.java

```diff
@@ -25,21 +31,77 @@ public static ProductRepository getInstance() {
         return INSTANCE;
     }
 
-    public void addProduct(Product product) {
-        int productId = idGenerator.getAndIncrement();
-        product.setId(productId);
-        productMap.put(product.getId(), product);
+    public Product addProduct(Product product) throws TheSameNameException, NoDataException, InvalidTypeOfDataException {
+        try {
+            if (product.getName() == null || product.getName().isEmpty()) {
+                throw new NoDataException("Product name is missing or empty.");
```

Comment:
przeczytaj sobie na spokojnie 1 i 2 zasadę SOLID.

Kod ten jest bardzo trudny do utrzymania - każde sprawdzenie wymaga dodawania nowego if'a.
A co jeśli Produkt będzie miał jeszcze wiecej pól ?

Musimy to zheremetyzować do innej metody, która będzie odpowiedzialna z walidacje produktu albo nawet do nowej klasy ProductValidator.
W ta metoda ma być tylko odpowiedzialna z dodanie produktu.